src/psbt/mod.rs

```diff
@@ -279,6 +280,37 @@ impl<'psbt, Pk: MiniscriptKey + ToPublicKey> Satisfier<Pk> for PsbtInputSatisfie
             <Satisfier<Pk>>::check_older(&Older(seq), n)
         }
     }
+
+    fn lookup_hash160(&self, h: hash160::Hash) -> Option<[u8; 32]> {
+        lookup_preimage(&self.psbt.inputs[self.index].hash160_preimages, &h)
+    }
+
+    fn lookup_sha256(&self, h: sha256::Hash) -> Option<[u8; 32]> {
+        lookup_preimage(&self.psbt.inputs[self.index].sha256_preimages, &h)
+    }
+
+    fn lookup_hash256(&self, h: sha256d::Hash) -> Option<[u8; 32]> {
+        lookup_preimage(&self.psbt.inputs[self.index].hash256_preimages, &h)
+    }
+
+    fn lookup_ripemd160(&self, h: ripemd160::Hash) -> Option<[u8; 32]> {
+        lookup_preimage(&self.psbt.inputs[self.index].ripemd160_preimages, &h)
+    }
+}
+
+// Lookup a [u8;32] preimage from a psbt hashmap
+// Returns Some() only if the preimage is 32 bytes
+fn lookup_preimage<T: Eq + Ord>(map: &BTreeMap<T, Vec<u8>>, key: &T) -> Option<[u8; 32]> {
```

Comment:
Not sure why we do not use here types available in the current PSBT implementation

Reply:
I don't follow this, Which type are you referring to? 

This is an internal function to avoid repetition of code in 4 satisfier definitions.

Reply:
What I am trying to say is that a hash preimage is a _type_. `[u8; 32]` is much more general (and used in many other circumstances than hash preimages), and a newtype can make a better job here — however it is not highly important if this is just a temporary stub.

Reply:
With the new BIP 174, we added fields for looking up preimages given hash. This is not a proprietary key. This function tells the Satisfier where to lookup for Preimages which it can use to construct satisfaction for generic Miniscripts. 

This is a generic lookup function cover the cases for `BTreeMap<sha256::Hash, Vec<u8>>`, `BTreeMap<sha256d::Hash, Vec<u8>>`, `BTreeMap<ripemd160::Hash, Vec<u8>>`, `BTreeMap<has160::Hash, Vec<u8>>`, into a single lookup function to avoid code repeatation.

Reply:
Ok, I was really confused b/c `map` in terms of PSBT/BIP 174 is either global, input or output map — and not a "map" inside a specific key values from PSBT map. Also "key" is the PSBT map key and not a hash of some secret.

Reply:
Can we do

```suggestion
fn lookup_preimage<T: bitcoin::Hash>(hash_preimages: &BTreeMap<T, Vec<u8>>, hash: &T) -> Option<Preimage> {
```

with `type Preimage = [u8; 32]` at least to help the reader/user to distinguishing `[u8; 32]` preimage from hash of the same dimension?

However, if I am right in a way I understand this function, I do not get what it does aside from the fact of looking up value for a key in a really odd fashion...

Reply:
Yes, it does lookup in a really weird fashion. The reason for that is 
1) PSBT spec does not enforce len = 32 on preimage, it is a constraint of miniscript. 
2) We get a `&[u8]` slice from the PSBT lookup, which we need to convert to [u8;32] if it has 32 length. If there is a more idiomatic MSRV compatible way to do this, please let me know. 

> However, if I am right in a way I understand this function, I do not get what it does aside from looking up value for a key in a really odd fashion...

This code of creating the 32 byte array from `&[u8]` slice if was len=32 was repeated 4 functions, so I forced it into a function. 

Nice idea with the Preimage32 type, I will add to the trait definition.

Reply:
Add a commit changing `[u8; 32]` to `Preimage32` in the `Satisfier` trait and in the impl for `psbt::InputSatisfier` for more readability.

Reply:
Ok, now much clearer.

Well, there is a function to copy slice into array: https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice

So what I would do here is to get rid of this function and use pretty simple line for looking up hashmap + copying from slice right inline. Moving that single line into a function makes code more obscure and complicated.

Reply:
This is my take on the problem: https://github.com/sanket1729/rust-miniscript/pull/2/files

Seems like we still need a separate fn to avoid duplication, but it is rather tiny and trivial in this case

Reply:
Thanks, this is approach is much cleaner. Picked and squashed your commit and added as co-author.